Minor Docker Optimizations

johnwesley opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description
<!--- Describe your changes in detail -->
Provide a `Makefile` to streamline the overall Docker process.
This includes an image build, container deploy, and cleanup target.

Remove `virtualenv` activation from the image build process. Containers are ephemeral in nature and I don't think we need the extra layer of virtualization in this case, however, `algo` inherently depends on it; so we still need to install the package and set the `env` for deploy.  

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
This change is not required, but I believe the `Makefile`  adds a level of convenience for Docker users.
Removing `virtualenv` activation from the build process installs `ansible` to the standard `/usr/local/bin` path so it's always available. I've had issues with this in the past with algo not finding the `ansible-playbook` binary location. 
<!--- If it fixes an open issue, please link to the issue here. -->

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
This has only been tested with `macOS 10.14.3`, `docker 18+`, and `DigitalOcean`. I can provide further tests in other environments if the maintainers are are open to these changes. This change will also require some documentation additions.

I have deployed with both `docker-deploy` Makefile target and `docker run --cap-drop=all -it -v ${HOME}/VPNS:/data trailofbits/algo:latest`, pointing to a local copy for `config.cfg`.
<!--- Include details of your testing environment, tests ran to see how -->
<!--- your change affects other areas of the code, etc. -->
### Caveats
1. `./algo-showenv.sh: line 72: ./env/bin/python: cannot execute binary file: Exec format error` is reported by `algo` because we have circumvented this.

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [] Bug fix (non-breaking change which fixes an issue)
- [] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] I have read the **CONTRIBUTING** document.
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [] I have updated the documentation accordingly.
- [] I have added tests to cover my changes.
- [] All new and existing tests passed.

jackivanov commented on Dockerfile:
```diff
@@ -15,11 +15,11 @@ RUN mkdir -p /algo && mkdir -p /algo/configs
 RUN apk --no-cache add ${BUILD_PACKAGES} && \
     python -m pip --no-cache-dir install -U pip && \
     python -m pip --no-cache-dir install virtualenv && \
     python -m virtualenv env && \
```
This one is should be removed as well

jackivanov commented:
Let's leave travis-ci for another issue and just delete `python -m virtualenv env` from the docker file

johnwesley commented:
I've about finished up, just one outstanding issue with `update-users.sh`. Can I open another PR with my working branch? https://travis-ci.org/johnwesley/algo

And should we keep the `Makefile` in this branch?

jackivanov commented:
Oh, if you have it already, lets address it in this PR as well :) Yes, keep Makefile in the same branch

johnwesley commented:
Merged the branch I've been testing against travis-ci. 

Test `local-deploy.sh` appears to pass.

There is still an issue with `update-users.sh` to resolve.

```
TASK [Generate the SSH private key] ********************************************
fatal: [10.0.8.24]: FAILED! => {"changed": false, "msg": "the python pyOpenSSL module is required"}
```

johnwesley commented:
@jackivanov  All tests now pass.  https://travis-ci.org/johnwesley/algo

johnwesley commented:
@jackivanov have you had a chance to review? And can I move forward with documentation updates?

jackivanov commented:
@johnwesley After my recent changes you would probably need to rebase from the master, and you can update the docs as well

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/trailofbits/algo?pullRequest=1341) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/trailofbits/algo?pullRequest=1341) before we can accept your contribution.<br/>**8** out of **9** committers have signed the CLA.<br/><br/>:white_check_mark: bshiznit<br/>:white_check_mark: dguido<br/>:white_check_mark: jackivanov<br/>:white_check_mark: johnwesley<br/>:white_check_mark: davidemyers<br/>:white_check_mark: TC1977<br/>:white_check_mark: adamluk<br/>:white_check_mark: faf0<br/>:x: akerl<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/trailofbits/algo?pullRequest=1341) it.</sub>

johnwesley commented:
Rebased with upstream master.
https://travis-ci.org/johnwesley/algo

Moving on to documentation.

jackivanov commented:
Looks like the branch is messed up with non related commints now.

johnwesley commented:
Will reset branch with master, add `Makefile` changes, and updates testing in a separate branch.

johnwesley commented:
@jackivanov I've updated the branch.

jackivanov commented on Dockerfile:
```diff
@@ -15,11 +15,11 @@ RUN mkdir -p /algo && mkdir -p /algo/configs
-    python -m pip --no-cache-dir install virtualenv && \
-    python -m virtualenv env && \
-    source env/bin/activate && \
+    #python -m pip --no-cache-dir install virtualenv && \
```
probably needs to be enabled?

johnwesley replied:
Local and CI tests pass without `virtualenv`.

jackivanov replied:
Ah, yes, right, we agreed to run it without virtualenv. Could you then just remote those lines, please?

jackivanov commented on deploy-from-docker.md:
```diff
@@ -0,0 +1,79 @@
+# Docker Support
```
seems like you've overwritten the whole file somehow. Could you include only your changes?

johnwesley replied:
Of course. Sorry.

jackivanov commented on tests/local-deploy.sh:
```diff
@@ -6,7 +6,7 @@ DEPLOY_ARGS="provider=local server=$LXC_IP ssh_user=ubuntu endpoint=$LXC_IP appa
 if [ "${LXC_NAME}" == "docker" ]
 then
-  docker run -it -v $(pwd)/config.cfg:/algo/config.cfg -v ~/.ssh:/root/.ssh -v $(pwd)/configs:/algo/configs -e "DEPLOY_ARGS=${DEPLOY_ARGS}" travis/algo /bin/sh -c "chown -R 0:0 /root/.ssh && source env/bin/activate && ansible-playbook main.yml -e \"${DEPLOY_ARGS}\" --skip-tags apparmor"
+  make docker-test-local
```
`docker-test-local` seems to be missing in the Makefile

johnwesley replied:
All tests are up to date.

johnwesley commented:
@jackivanov I've (hopefully) saved this branch and addressed your concerns. Tests have been updated to use the `Makefile` and the original doc has been restored.

jackivanov commented:
My tests failed:
```
root@ip-172-26-13-231:~/docker-optimize# make docker-deploy
docker run \
--cap-drop=all \
--rm \
-it \
-v /home/ubuntu/docker-optimize:/data \
trailofbits/algo:latest
/algo/env/bin/activate not found.  Did you follow documentation to install dependencies?
Makefile:20: recipe for target 'docker-deploy' failed
make: *** [docker-deploy] Error 1
```

I guess we need to modify `./algo` to skip this check if deploying from docker

johnwesley commented:
This is my local play

```
$ make docker-deploy
docker run \
        --cap-drop=all \
        --rm \
        -it \
        -v /Users/kwisatz4/Github/algo:/data \
        trailofbits/algo:latest

PLAY [localhost] ******************************************************************************************************

TASK [Gathering Facts] *******************************************************************************************************
ok: [localhost]

TASK [Verify Ansible meets Drupal VM's version requirements.] *******************************************************************************************************
ok: [localhost] => {
    "changed": false,
    "msg": "All assertions passed"
}

PLAY [Ask user for the input] 
```
Will try from different environments.

johnwesley commented:
And this is my local play (Arch)
```
[vagrant@archlinux algo]$ sudo make docker-deploy                                                                                                                                    │
docker run \                                                                                                                                                                         │
--cap-drop=all \                                                                                                                                                                     │
--rm \                                                                                                                                                                               │
-it \                                                                                                                                                                                │
-v /home/vagrant/algo:/data \                                                                                                                                                        │
trailofbits/algo:latest                                                                                                                                                              │
/algo/env/bin/activate not found.  Did you follow documentation to install dependencies?                                                                                             │
make: *** [Makefile:20: docker-deploy] Error 1
```

johnwesley commented:
@jackivanov testing a modification to `./algo` in clean Linux and OSX environments.

johnwesley commented:
Tested the modified condition with an Ubuntu 18 Vagrant environment. Both docker and standard deploys kicked off.

dguido commented:
What do we need to get this merged?

johnwesley commented:
Sorry, I was waiting on your feedback after my last update.
I would suggest we resolve the conflicts and the the `update-users` test. Did you have a chance to review the changes to the main `algo` script?

jackivanov commented:
The fix in `algo` looks good to me. Could you, please, fix the conflicts and I'll take it over from there?

johnwesley commented:
Yes I will.

TC1977 commented on docs/deploy-from-docker.md:
```diff
@@ -50,6 +50,15 @@ $ docker run --cap-drop=all -it \
+**This has not been tested in a Windows environment.**
+
+1. From the project's root run `make docker-build`. This builds a Docker image labeled `trailofbits/algo:latest`. These values can be changed in the `Makefile` itself.
+2. Run `make docker-deploy`. By default this will use the given copy of `config.cfg` and populate the project's root `~/configs/` directory with your client configuration data. If you would rather use a local directory and `config.cfg` file, then follow steps 2 and 3 before changing the `CONFIGURATIONS` variable in the `Makefile`.
```
Please clarify the "steps 2 and 3" that are referred to here?

johnwesley replied:
Yes, I will expand on this.

johnwesley commented:
@jackivanov conflicts have been resolved and the first local test `docker-ci-local` passes. Upgrade users test is failing on a step.

johnwesley commented:
@jackivanov I've also tested `algo` locally with both Docker and `virtualenv` deployments.

jackivanov commented:
Do we really need to mark the image as `trailofbits/algo:latest`? Won't we get conflicts with [the one from dockerhub](https://cloud.docker.com/u/trailofbits/repository/docker/trailofbits/algo) in some cases?

johnwesley commented:
No we don't. Would you rather use a release tag of some sort?
We can use `$(IMAGE):$(TAG)` in the test runs and set the tag top-scope same as we're doing for `docker-build` and `docker-deploy`. Another option is to use a git tag.
Example:
`TAG := $(shell git describe --tags)`

`Successfully tagged trailofbits/algo:v1.0-65-g91a44bf`

jackivanov commented:
I think we're good after you complete [this](https://github.com/trailofbits/algo/pull/1341/files#r290357912)

johnwesley commented:
Sorry for this, but I think attempting to remove Python virtualenv from the docker deploy is a bad idea at this point and will cause systemic problems down the line. It's too ingrained into the app and I'm not 100% confident this will work for every situation presented. I say we move forward with the `Makefile` for Docker and tackle `virtualenv` in another PR.

jackivanov commented:
@johnwesley I agree. Could you revert the changes, please?

jackivanov commented:
The tests are failing now @johnwesley

johnwesley commented:
@jackivanov I am aware and will try to fix resolve them today.

johnwesley commented:
@jackivanov Docs have been updated and tests pass.

johnwesley commented:
Can we just close this PR so I can open a new one with just the `Makefile` addition for Docker and tests?
